pkg/guacanalytics/patchPlanning.go

```diff
@@ -0,0 +1,187 @@
+		return nil, fmt.Errorf("failed getting intial node with given ID:%w", err)
+	}
+
+	nodePkg, ok := startNode.Node.(*model.NodeNodePackage)
+
+	if !ok {
+		return nil, fmt.Errorf("Not a package")
+	}
+
+	q := queueValues{
+		queue:   make([]string, 0), // the queue of nodes in bfs
+		nodeMap: map[string]DfsNode{},
+	}
+
+	// TODO: add functionality to start with other nodes?
+	if len(nodePkg.AllPkgTree.Namespaces) < 1 {
+		return nil, fmt.Errorf("Start by inputting a packageName or packageVerion node")
+	}
+
+	if len(nodePkg.AllPkgTree.Namespaces[0].Names) < 1 {
+		return nil, fmt.Errorf("Start by inputting a packageName or packageVerion node")
+	}
+
+	if len(nodePkg.AllPkgTree.Namespaces[0].Names[0].Versions) < 1 {
+		var versionsList []string
+		for _, versionEntry := range nodePkg.AllPkgTree.Namespaces[0].Names[0].Versions {
+			versionsList = append(versionsList, versionEntry.Version)
+		}
+		q.nodeMap[startID] = DfsNode{
+			NodeType:     "packageName",
+			nodeVersions: versionsList,
+		}
+	} else {
+		q.nodeMap[startID] = DfsNode{
+			NodeType: "packageVersion",
+		}
+
+		// add packageName node to the frontier as well
+		q.queue = append(q.queue, nodePkg.AllPkgTree.Namespaces[0].Names[0].Id)
+
+		var versionsList []string
+		versionsList = append(versionsList, nodePkg.AllPkgTree.Namespaces[0].Names[0].Versions[0].Version)
+		q.nodeMap[nodePkg.AllPkgTree.Namespaces[0].Names[0].Id] = DfsNode{
+			NodeType:     "packageName",
+			nodeVersions: versionsList,
+		}
+	}
+
+	q.queue = append(q.queue, startID)
+
+	for len(q.queue) > 0 {
+		q.now = q.queue[0]
+		q.queue = q.queue[1:]
+		q.nowNode = q.nodeMap[q.now]
+
+		if stopID != nil && *stopID == q.now {
+			break
+		}
+
+		if q.nowNode.Depth >= maxDepth {
+			break
+		}
+
+		neighborsResponse, err := model.Neighbors(ctx, gqlclient, q.now, []model.Edge{})
+
+		if err != nil {
+			return nil, fmt.Errorf("failed getting neighbors:%w", err)
+		}
+
+		for _, neighbor := range neighborsResponse.Neighbors {
+			err = caseOnPredicates(ctx, gqlclient, &q, neighbor, q.nowNode.NodeType)
+
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		q.nowNode.Expanded = true
+		q.nodeMap[q.now] = q.nowNode
+	}
+
+	return q.nodeMap, nil
+
+}
+
+func caseOnPredicates(ctx context.Context, gqlclient graphql.Client, q *queueValues, neighbor model.NeighborsNeighborsNode, nodeType string) error {
+	// case on predicates and nodeType
+	switch nodeType {
+	case "packageName":
+		switch neighbor := neighbor.(type) {
+		case *model.NeighborsNeighborsIsDependency:
+			err := exploreIsDependency(ctx, gqlclient, q, *neighbor)
+
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func exploreIsDependency(ctx context.Context, gqlclient graphql.Client, q *queueValues, isDependency model.NeighborsNeighborsIsDependency) error {
+	doesRangeInclude, err := depversion.DoesRangeInclude(q.nowNode.nodeVersions, isDependency.VersionRange)
+
+	if err != nil {
+		return err
+	}
+	if doesRangeInclude {
```

Comment:
nit: to make it more readable, you can do the inverse for block which include everything. So for this, something like

```
if !doesRangeInclude {
    return nil
}

// do other stuff assuming doesRangeInclude==true
```